### ISSUE TITLE
CA1812 should not warn for top-level Program classes

### ISSUE THREAD

**Issue by martincostello** [dotnet/roslyn-analyzers#5628]
### Analyzer

**Diagnostic ID**: [CA1812](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1812)

### Describe the improvement

An application that uses top-level programs and uses an analysis rule set that includes CA1812 will generate a warning from the compiler-generated `Program` class in even a trivial application.

```csharp
Console.WriteLine("Hello, World!");
```

```
Program.cs(1,1): warning CA1812: Program is an internal class that is apparently never instantiated. If so, remove the code from the assembly. If this class is intended to contain only static members, make it static (Shared in Visual Basic). [Project.csproj]
```

This then requires the developer to suppress it like so:

```csharp
#pragma warning disable CA1812
Console.WriteLine("Hello, World!");
```

### Describe suggestions on how to achieve the rule

Special-case compiler-generated `Program` classes for top-level programs and do not emit the violation.

### Additional context

Application that reproduces the problem adapted from `dotnet new console` using the .NET 6 RC2 SDK.

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <AnalysisMode>AllEnabledByDefault</AnalysisMode>
    <EnableNETAnalyzers>true</EnableNETAnalyzers>
    <ImplicitUsings>enable</ImplicitUsings>
    <NoWarn>$(NoWarn);CA1014;CA1303</NoWarn>
    <Nullable>enable</Nullable>
    <OutputType>Exe</OutputType>
    <TargetFramework>net6.0</TargetFramework>
  </PropertyGroup>
</Project>
```

```csharp
Console.WriteLine("Hello, World!");
```

**Comment by Youssef1313** [comment]
@martincostello Can you try with https://dev.azure.com/dnceng/public/_packaging?_a=package&feed=dotnet7&package=Microsoft.CodeAnalysis.NetAnalyzers&version=7.0.0-preview1.21513.1&protocolType=NuGet ?

**Comment by adamralph** [comment]
This still occurs in the latest RC (6.0.100-rc.2.21505.57), see https://github.com/adamralph/minver/runs/4132244037?check_suite_focus=true

Weirdly, it's only a problem on Windows and MacOS, but not Linux.

Is the intention to fix this before RTM?

**Comment by Youssef1313** [comment]
It looks like the milestone for https://github.com/dotnet/roslyn-analyzers/pull/5351 was set to .NET 7.

@mavasani @jmarolf Can this get in .NET 6?

**Comment by martincostello** [comment]
Yeah I've seen this not fixed in 6.0.0/6.0.100 either, but that's already been built and ships today so would need backporting to 6.0.1.

**Comment by adamralph** [comment]
@martincostello I realise RTM is tomorrow, but all I see at https://dotnet.microsoft.com/download/dotnet/6.0 is 6.0.100-rc.2.21505.57.

**Comment by martincostello** [comment]
I _believe_ it's shipping today ahead of .NET Conf tomorrow - the 6.0.0 packages are all in NuGet.org now (I woke up this morning to various dependabot updates), and the `6.0.100` SDK has been present at the URL it lives at for CI scripts for a week or so. They just haven't updated the [`releases.json`](https://github.com/dotnet/core/blob/main/release-notes/6.0/releases.json) file (which powers the .NET website's download links) yet.

**Comment by jmarolf** [comment]
> It looks like the milestone for #5351 was set to .NET 7.
> 
> Manish Vasani Jonathon Marolf Can this get in .NET 6?

sadly, no by the time the fix was merged final builds of .NET 6 were already spinning for today. Considering CA1812 is off by default I don't think this makes the bar to port as a day 0 fix. but I will pursue it for the 6.0.200 SDK release

**Comment by michha** [comment]
@jmarolf
Will this be fixed in 6.0.200 SDK? Is there a rough timeline for its release?

**Comment by Grauenwolf** [comment]
It is fixed in 7.0.0-preview1.22168.1.

```
	<ItemGroup>
		<PackageReference Include="Microsoft.CodeAnalysis.NetAnalyzers" Version="7.0.0-preview1.22168.1">
			<IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
			<PrivateAssets>all</PrivateAssets>
		</PackageReference>
	</ItemGroup>
```

**Comment by Youssef1313** [comment]
@jmarolf @mavasani This one is already fixed, but for 7.x.

The request here for this issue is backporting to 6.x. I don't know if it meets the bar for backporting. If not, can we close?

**Comment by MrDaedra** [comment]
Any updates on this fix?

@Youssef1313 as .NET 6 is in LTS it makes sense to backport it

**Comment by mavasani** [comment]
Tagging @buyaa-n

**Comment by buyaa-n** [comment]
As @jmarolf mentioned the analyzer is off by default and there is a workaround (suppress), therefore this does not meet the bar for servicing 6.0, I guess it could be considered if 1st party customer requesting it

**Comment by BryceBarbara** [comment]
Just to share with others who might be countering this, I was able to disable this for all projects in my solution by putting this in the `.editorconfig`

```editorconfig
# Workaround for https://github.com/dotnet/roslyn-analyzers/issues/5628
[Program.cs]
dotnet_diagnostic.ca1812.severity = none
```

**Comment by VanessaRussell** [comment]
> Just to share with others who might be countering this, I was able to disable this for all projects in my solution by putting this in the `.editorconfig`
> 
> ```editorconfig
> # Workaround for https://github.com/dotnet/roslyn-analyzers/issues/5628
> [Program.cs]
> dotnet_diagnostic.ca1812.severity = none
> ```

This does not work for me 😢

```
Microsoft Visual Studio Enterprise 2022
Version 17.2.6
VisualStudio.17.Release/17.2.6+32630.192
Microsoft .NET Framework
Version 4.8.04161

Installed Version: Enterprise

.NET Core Debugging with WSL   1.0
.NET Core Debugging with WSL

ADL Tools Service Provider   1.0
This package contains services used by Data Lake tools

ASA Service Provider   1.0

ASP.NET and Web Tools 2019   17.2.393.26812
ASP.NET and Web Tools 2019

Azure App Service Tools v3.0.0   17.2.393.26812
Azure App Service Tools v3.0.0

Azure Data Lake Tools for Visual Studio   2.6.5000.0
Microsoft Azure Data Lake Tools for Visual Studio

Azure Functions and Web Jobs Tools   17.2.393.26812
Azure Functions and Web Jobs Tools

Azure Stream Analytics Tools for Visual Studio   2.6.5000.0
Microsoft Azure Stream Analytics Tools for Visual Studio

C# Tools   4.2.0-4.22281.5+8d3180e5f00d42f0f0295165f756f368f0cbfa44
C# components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Common Azure Tools   1.10
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

File Icons   2.7
Adds icons for files that are not recognized by Solution Explorer

IndentRainbowPackage Extension   1.0
IndentRainbowPackage Visual Studio Extension Detailed Info

Microsoft Azure Hive Query Language Service   2.6.5000.0
Language service for Hive query

Microsoft Azure Stream Analytics Language Service   2.6.5000.0
Language service for Azure Stream Analytics

Microsoft Azure Tools for Visual Studio   2.9
Support for Azure Cloud Services projects

Microsoft JVM Debugger   1.0
Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

Node.js Tools   1.5.40404.3 Commit Hash:dd524a7d0ee653de8b7b32a93ca22f2d57cd2419
Adds support for developing and debugging Node.js apps in Visual Studio

NuGet Package Manager   6.2.1
NuGet Package Manager in Visual Studio. For more information about NuGet, visit https://docs.nuget.org/

Razor (ASP.NET Core)   17.0.0.2218101+885a343b00bcab620a90c1550c37dafd730ce984
Provides languages services for ASP.NET Core Razor.

SQL Server Data Tools   17.0.62204.01010
Microsoft SQL Server Data Tools

Syntax Visualizer   1.0
An extension for visualizing Roslyn SyntaxTrees.

ToolWindowHostedEditor   1.0
Hosting json editor into a tool window

TypeScript Tools   17.0.10418.2001
TypeScript Tools for Microsoft Visual Studio

Visual Basic Tools   4.2.0-4.22281.5+8d3180e5f00d42f0f0295165f756f368f0cbfa44
Visual Basic components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Visual F# Tools   17.1.0-beta.22329.1+702b8e77f5fbfe21e6743324c1750503e02f182d
Microsoft Visual F# Tools

Visual Studio Color Theme Designer   1.0
Allows users to create and edit Visual Studio themes. Themes can customize colors used for menus, toolbars, tabs, titlebars, the text editor, and other environment colors.

Visual Studio IntelliCode   2.2
AI-assisted development for Visual Studio.

VSColorOutput64   2022.2
Color output for build and debug windows - https://mike-ward.net/vscoloroutput
```

**Comment by Youssef1313** [comment]
Closing as this is fixed in 7.x and doesn't meet the bar to backport for 6.x

**Comment by kimsey0** [comment]
Which version is this fixed from? I'm still getting a warning if I create a new console project and enable the `Recommended` analysis mode with the .NET SDK version 7.0.101.

**Comment by Youssef1313** [comment]
@kimsey0 I think the fix didn't yet flow into the .NET SDK.

Can you try with 7.0.0 from NuGet.org?

**Comment by kimsey0** [comment]
That does indeed seem to fix it. I just assumed that since Microsoft.CodeAnalysis.NetAnalyzers 7.0.0 was released on December 8th and .NET SDK 7.0.1 was released on December 13th, it would include the newest analyzers, but there might be a bit of a delay for the flow. (7.0.1 is also marked as a security patch, so I don't know if that influences anything.)

I hope the analyzer changes will be out in the next SDK release. :-)
